File: src/main/java/uk/gov/justice/digital/pdf/Configuration.java
package uk.gov.justice.digital.pdf;

import com.google.common.collect.ImmutableMap;
import com.google.inject.AbstractModule;
import com.google.inject.Binder;
import java.util.Map;
import java.util.function.Function;
import uk.gov.justice.digital.pdf.interfaces.TemplateRepository;
import uk.gov.justice.digital.pdf.mixin.EnvironmentBinder;
import uk.gov.justice.digital.pdf.service.ResourceRepository;

public class Configuration extends AbstractModule implements EnvironmentBinder {

    @Override
    public Map<String, String> envDefaults() {

        return ImmutableMap.of(
                "PORT", "8080",
                "DEBUG_LOG", "false",
                "ALFRESCO_URL", "http://localhost:8080/alfresco/service/",
                "ALFRESCO_USER", "alfrescoUser"
        );
    }

    @Override
    protected final void configure() {

        bindConfiguration(
                Integer.class,
                Integer::parseInt,
                ImmutableMap.of(
                        "port", "PORT"
                )
        );

        bindConfiguration(
                Boolean.class,
                Boolean::parseBoolean,
                ImmutableMap.of(
                        "debugLog", "DEBUG_LOG"
                )
        );

        bindConfiguration(
                String.class,
                Function.identity(),
                ImmutableMap.of(
                        "alfrescoUrl", "ALFRESCO_URL",
                        "alfrescoUser", "ALFRESCO_USER"
                )
        );

        configureOverridable();
    }

    protected void configureOverridable() {

        bind(TemplateRepository.class).to(ResourceRepository.class);
    }

    @Override
    public final Binder binder() {

        return super.binder();  // Mix in binder() from AbstractModule to EnvironmentBinder
    }
}

File: src/main/kotlin/uk/gov/justice/digital/pdf/Configuration.kt
package uk.gov.justice.digital.pdf

import com.google.inject.AbstractModule
import uk.gov.justice.digital.pdf.interfaces.TemplateRepository
import uk.gov.justice.digital.pdf.mixins.EnvironmentBinder
import uk.gov.justice.digital.pdf.service.ResourceRepository

open class Configuration : AbstractModule(), EnvironmentBinder {

    override val envDefaults = mapOf(
            "PORT" to "8080",
            "DEBUG_LOG" to  "false",
            "ALFRESCO_URL" to  "http://localhost:8080/alfresco/service/",
            "ALFRESCO_USER" to "alfrescoUser"
    )

    override fun binder() = super.binder()

    override fun configure() {

        bindConfiguration({ it?.toInt() }, mapOf("port" to "PORT"))
        bindConfiguration({ it?.toBoolean() }, mapOf("debugLog" to "DEBUG_LOG"))
        bindConfiguration({ it }, mapOf(
                "alfrescoUrl" to "ALFRESCO_URL",
                "alfrescoUser" to "ALFRESCO_USER"
        ))

        configureOverridable()
    }

    fun configureOverridable() {

        bind(TemplateRepository::class.java).to(ResourceRepository::class.java)
    }

    private inline fun <reified T> bindConfiguration(noinline transform: (String?) -> T, map: Map<String, String>) =

        bindConfiguration(T::class.java, transform, map)
}
